Lib/importlib/_bootstrap.py

```diff
@@ -957,16 +959,16 @@ def _find_and_load_unlocked(name, import_):
 
 def _find_and_load(name, import_):
     """Find and load the module."""
-    _imp.acquire_lock()
-    if name not in sys.modules:
-        with _ModuleLockManager(name):
+    with _ModuleLockManager(name):
+        if name not in sys.modules:
             return _find_and_load_unlocked(name, import_)
+
     module = sys.modules[name]
     if module is None:
-        _imp.release_lock()
         message = ('import of {} halted; '
                    'None in sys.modules'.format(name))
         raise ModuleNotFoundError(message, name=name)
+
```

Comment:
Even with your improvements to the lock handling, this still looks a bit race-prone to me, since we have the classic "query before use" pattern of:

    if name not in sys.modules:
        ...
    module = sys.modules[name]

That is, just because the module was there when we checked `if name not in sys.modules` doesn't mean it's *still* going to be there when we run `module = sys.modules[name]`.

Previously, holding `_imp.acquire_lock()` for the whole function would at least protect this from other `_find_and_load()` calls, but as far as I can see it's *never* been protected from other threads doing `del sys.modules[name]` without holding the relevant module lock.